### ISSUE TITLE
Illegal ctor redirect raises NPE instead of a useful error

### ISSUE THREAD

**Issue by Runemoro** [SpongePowered/Mixin#270]
Exception: https://paste.dimdev.org/yilaqohabe.txt
Mixin:
```java
    @Redirect(method = "<init>(Lnet/minecraft/block/Block;[Lnet/minecraft/block/properties/IProperty;)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/state/BlockStateContainer;<init>(Lnet/minecraft/block/Block;[Lnet/minecraft/block/properties/IProperty;Lcom/google/common/collect/ImmutableMap;)V"))
    public void init(Block block, IProperty<?>[] properties, ImmutableMap<IUnlistedProperty<?>, Optional<?>> unlistedProperties) {
        this.block = block;
        ImmutableSortedMap.Builder<String, IProperty<?>> propertiesBuilder = ImmutableSortedMap.naturalOrder();

        for (IProperty<?> property : properties) {
            validateProperty(block, property);
            propertiesBuilder.put(property.getName(), property);
        }

        this.properties = propertiesBuilder.build();
    }
```
It seems like `superCall` is null when calling `target.indexOf(superCall)` in `InvokeInjector`.

**Comment by Mumfrey** [comment]
You can't redirect a constructor like this, read the documentation. However it should at least provide a useful error message instead of the NPE, so that part is valid.

**Comment by Mumfrey** [comment]
I can't even see where in the code this is mixing in to. I cannot find an invocation of `BlockStateContainer`'s ctor inside another ctor anywhere in vanilla minecraft. Since you failed to actually provide the parent mixin I have no idea how I'm supposed to recreate your test case.

**Comment by Mumfrey** [comment]
Wait, **what**?!?!!?!

The initial example is **trying to redirect a SUPER CONSTRUCTOR CALL**?

I can't even.







....

Still doesn't explain the actual bug but seriously this is some dumb crap.

**Comment by Runemoro** [comment]
@gabizou @Mumfrey I am not redirecting the super constructor. The mixin's target is `BlockStateContainer`, not `ExtendedBlockStateContainer` (which both have the same constructor).

The inability to modify constructors is a major flaw in Mixin, in my opinion. It's forcing me to use ASM directly to patch some of the constructors I'm not able to edit using Mixin.

**Comment by Mumfrey** [comment]
Ok so my superctor lookup is failing because this is actually an overloaded ctor. Makes sense, that can be fixed.

> The inability to modify constructors is a major flaw in Mixin, in my opinion.

You're entitled to your opinion, but I directly reject your assertion. **The inability to modify constructors is a deliberate feature of Mixin.**

Consider the things that you **can** do with Mixin to interact with object construction:

* Use modifyargs and modifyarg to modify ctor args inline
* Redirect any field or method call in a ctor, just like other methods
* Modify constants ("redirect" them if you will) in a ctor
* Modify any initialiser and override initialisers with your own
* Append callbacks to the end of ctors for custom construction logic.

Now consider the things I don't let you do:

* Inject at arbitrary points in the ctor - ctors are delicate enough as it is, this is pure sanity stuff
* Redirect the a ctor call - this is just utterly stupid because ctor calls should **never** be treated like a normal method call

> It's forcing me to use ASM directly to patch some of the constructors I'm not able to edit using Mixin.

Doing stupid shit with ASM is your prerogative, but I'm not going to endorse or support it. If you can't do what you need to with the tools in Mixin then you probably shouldn't be doing it. I don't even see anything in your example that you couldn't do without your proposed solution of redirecting the ctor anyway. If you just want to modify the setting of `properties` then do so by redirecting the field setter.

I accept that the bug here is that I didn't take into account delegating (overloaded) ctors, and will fix it accordingly. However redirecting constructors is reckless and 100% unnecessary, I wouldn't recommend doing it in ASM and it's certainly never going to happen in Mixin.

**Comment by Runemoro** [comment]
>  If you can't do what you need to with the tools in Mixin then you probably shouldn't be doing it.

The only reason I can't do what I need to do is that Mojang decided to put the initialization code for the `BlockStateContainer` in the constructor. If they had simply put it in a separate method named `init` which was called by the constructor, I would be able to inject into it, but just because the method I want to inject into is named `<init>` rather than `init`, Mixin refuses to do so.

> ctor calls should never be treated like a normal method call

But they are. Once compiled to bytecode, a constructor is just a method named `<init>` that happens to be called after every object creation. The following code:
```
public class TestClass extends SuperClass {
    private int x;
    private int y;
    private String s1 = "String 1";
    private String s2;

    {
        s2 = "String 2";
    }

    public TestClass(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public void init(int x, int y) {
        super.init();
        s1 = "String 1";
        s2 = "String 2";
        this.x = x;
        this.y = y;
    }
}
```

compiles to the following bytecode:

```java
  public void <init>(int x, int y);
    Flags: PUBLIC
    Code:
      stack=2, locals=3, arguments=2
           0: aload_0         /* this */
           1: invokespecial   SuperClass.<init>:()V
           4: aload_0         /* this */
           5: ldc             "String 1"
           7: putfield        TestClass.s1:Ljava/lang/String;
          10: aload_0         /* this */
          11: ldc             "String 2"
          13: putfield        TestClass.s2:Ljava/lang/String;
          16: aload_0         /* this */
          17: iload_1         /* x */
          18: putfield        TestClass.x:I
          21: aload_0         /* this */
          22: iload_2         /* y */
          23: putfield        TestClass.y:I
          26: return         
    LocalVariableTable:
      Start  Length  Slot  Name  Signature
      -----  ------  ----  ----  -----------
      0      27      0     this  LTestClass;
      0      27      1     x     I
      0      27      2     y     I
  
  public void init(int x, int y);
    Flags: PUBLIC
    Code:
      stack=2, locals=3, arguments=2
           0: aload_0         /* this */
           1: invokespecial   SuperClass.init:()V
           4: aload_0         /* this */
           5: ldc             "String 1"
           7: putfield        TestClass.s1:Ljava/lang/String;
          10: aload_0         /* this */
          11: ldc             "String 2"
          13: putfield        TestClass.s2:Ljava/lang/String;
          16: aload_0         /* this */
          17: iload_1         /* x */
          18: putfield        TestClass.x:I
          21: aload_0         /* this */
          22: iload_2         /* y */
          23: putfield        TestClass.y:I
          26: return         
    LocalVariableTable:
      Start  Length  Slot  Name  Signature
      -----  ------  ----  ----  -----------
      0      27      0     this  LTestClass;
      0      27      1     x     I
      0      27      2     y     I
  ```

The methods are **exactly** the same, apart from one of them being named `<init>` rather than `init`.

Here are the possible "problems" when injecting into/overwriting a constructor:

 1. It's difficult to know what instruction the injection point will target exactly because the compiled constructor includes field and initializers at the top. -> Yes, it's a bit more difficult than a regular method, but the order of the field initializer is preserved, making it much easier than something like local variable capture, which mixin supports.
2. An overwrite might leave a final field uninitialized -> Mixin could check this by making sure all `@Final` fields are assigned to in the overwriting method.

**Comment by DemonWav** [comment]
> but just because the method I want to inject into is named <init> rather than init, Mixin refuses to do so.

This sentence screams such ignorance it's frightening.

**Comment by Runemoro** [comment]
@DemonWav 

        if (Constants.CTOR.equals(target.method.name)) { // Constants.CTOR is "<init>"
            for (InjectionPoint injectionPoint : injectionPoints) {
                if (!injectionPoint.getClass().equals(BeforeReturn.class)) {
                    throw new InvalidInjectionException(this.info, "Found injection point type " + injectionPoint.getClass().getSimpleName()
                            + " targetting a ctor in " + this + ". Only RETURN allowed for a ctor target");
                }
            }
        }

Removing that check makes it work, I tested it.

**Comment by DemonWav** [comment]
[And here is a link that shows you how to reflectively assign a new value to a `private static final` variable.](https://stackoverflow.com/a/3301720/2353230)

It's almost as if "able to do something" and "should do something" and "will always work" are 3 entirely separate categories of things.

**Comment by Runemoro** [comment]
> Even just the fact that ctors are called by invokespecial rather than invokevirtual
And how would that affect injecting into a constructor?

> And here is a link that shows you how to reflectively assign a new value to a private static final variable.
What does that have anything to do with this?

**Comment by DemonWav** [comment]
> And how would that affect injecting into a constructor?

It doesn't, simply making a point.

> How does that have anything to do with this?

Just making a point, as my second line says.

> You say that, and yet you haven't been able to give a single example of a problem that might be caused by injecting into a constructor.

Field initialization guarantees, field initialization ordering, super and primary constructor calls, etc. There are concrete examples (and cases I'm not aware of) that I'm sure will be provided, but as this isn't a project I work on, don't have any at hand.

**Comment by Runemoro** [comment]
> Field initialization guarantees
As I stated in my previous comment, Mixin could check that an overwrite/injection into a constructor initializes all final fields once and only once.

> field initialization ordering
As I stated in my previous comment, this is not a problem since the constructor always initializes the fields in the class. But some parts of Mixin such as local variable capture require you to know the bytecode of the target method, so determining an injection point for the constructor is much easier compared to this.

> super and primary constructor calls

The mixin class should extend the target's super class. The overwritten constructor will necessarily include a super call to a correct constructor.

**Comment by Pokechu22** [comment]
> the Mixin contract is that the Mixin will extend some class within the class hierarchy of the target, up to and including Object.

Not just that; they can also extend [other mixins in the hierarchy](https://github.com/SpongePowered/Mixin/wiki/About-Hierarchy-Validation-in-Mixins#mixin-inheritance) too.  You won't have any meaningful constructor in that case.

**Comment by Runemoro** [comment]
If you want to use the super constructor of a package-private class, then you can just extend a mixin targeting it which `@Shadow`s the constructor (same as if you wanted to access any other super method).

If you want to overwrite a constructor and extend a mixin, you would have to add an intermediate mixin targeting the super class of your mixin's target, add a shadow constructor to that one, and make that one extend the other mixin class.

Mixins overwriting constructors with multiple targets would still be supported as long as they all extend the same class.

**Comment by Mumfrey** [comment]
I don't know why you guys are arguing about this, I thought I made my position clear above.

> Removing that check makes it work, I tested it.

Of course it does. And "proving" that a ctor is a normal method by showing how they're normal methods in the bytecode. Did it occur to you that I know that and that my statement *ctor calls should never be treated like a normal method call* is about how they're *treated* not what they *are*? Did it not occur to you that maybe having written a bytecode-engineering library as large as Mixin that I might already know that?

> I am trying to completely remove this from the constructor, and only create instances of StateImplementation as they are needed, to optimize memory usage:

Well firstly that sounds like something which should be PR'd to Forge, then it's in the base implementation that a lot of mods are using and it will be clearer to them that lazy-loading is occurring. This is especially important given that lazy-loading changes the contract of the class somewhat. But even so, I can think of several ways to inhibit the initialisation *without* nuking the constructor from orbit. Consider:

```java
@Mixin(BlockStateContainer.class)
public abstract class MixinBlockStateContainer
{
    private static final String TARGET_CTOR = "<init>(Lnet/minecraft/block/Block;[Lnet/minecraft/block/properties/IProperty;Lcom/google/common/collect/ImmutableMap;)V";
    private static final String CARTESIAN_PRODUCT = "Lnet/minecraft/util/math/Cartesian;cartesianProduct(Ljava/lang/Iterable;)Ljava/lang/Iterable;";
    private static final String NEW_LINKED_HASHMAP = "Lcom/google/common/collect/Maps;newLinkedHashMap()Ljava/util/LinkedHashMap;";

    // Returning an empty list here defeats the population logic for the state
    // implementation map and ensures that both the map (which we null out to
    // avoid allocation) and list which becomes the validStates member are never
    // populated
    @Redirect(method = TARGET_CTOR, at = @At(value = "INVOKE", target = CARTESIAN_PRODUCT))
    private <T> Iterable<List<T>> cartesianProduct(Iterable <? extends Iterable <? extends T >> sets) {
        return Collections.<List<T>>emptyList();
    }

    // This map is never used since we defeat the call to cartesianProduct.
    @Redirect(method = TARGET_CTOR, at = @At(value = "INVOKE", target = NEW_LINKED_HASHMAP))
    private LinkedHashMap<?, ?> newLinkedHashMap()
    {
        return null;
    }
}
```

Which demonstrates that simply defeating the call to `cartesianProduct` actually has a cascade effect of basically not initialising anything else. Hitting one small "pressure point" in the method's logic keeps the method intact but defeats its functionality. Returning the `null` map isn't strictly necessary but means the allocation is inhibited. A further tweak could be to add a redirect for the `newArrayList` call to return an `ArrayList` with built-in lazy-loading on `validStates`:

```java
private static final String NEW_ARRAYLIST = "Lcom/google/common/collect/Lists;newArrayList()Ljava/util/ArrayList;";

@Redirect(method = TARGET_CTOR, at = @At(value = "INVOKE", target = NEW_ARRAYLIST))
private <E> ArrayList<E> newArrayList() {
   return new CustomLazyLoadingArrayList<E>();
}
```

As I said above. I intend to fix the **valid** bug that this injection should have triggered a proper error message instead of a cryptic NPE. I will not add any kind of support for modifying constructors.

**Comment by Mumfrey** [comment]
Closed by 45e152690c2114475d212b37651e41c86e518424
